RFC0032.md

```diff
@@ -0,0 +1,112 @@
+
+---
+
+**Advice to users:**  
+The available channel support on a system can be queried via the PMIx\_Query\_nb API should the application developer wish to tailor their code accordingly - this will always report the channels directly supported by the PMIx library. However, channels supported by the host RM will be included only if the RM itself supports such queries.
+
+The PMIx\_Log API should _never_ be used for streaming data as it is not a "performant" transport and can perturb the application since it involves the local PMIx server and host RM daemon.
+
+There is some ambiguity regarding what information is provided in the "data" parameter vs the "directives" parameter when calling PMIx\_Log. In general, information that is to be included in the log should be provided in the "data" array. Examples would include:
+```c
+PMIX_INFO_LOAD(&data[0], PMIX_LOG_STDOUT, "my message string", PMIX_STRING);
+PMIX_INFO_LOAD(&data[1], PMIX_LOG_STDERR, "my error message", PMIX_STRING);
+```
+Note that in the above examples, the attribute key defines the channel to be used for the provided data.
+
+On the other hand, "directives" should be used to request behaviors of the PMIx\_Log function. Examples might include:
+```c
+PMIX_INFO_LOAD(&directives[0], PMIX_LOG_GENERATE_TIMESTAMP, NULL, PMIX_BOOL);
+PMIX_INFO_LOAD(&directives[1], PMIX_LOG_XML_OUTPUT, NULL, PMIX_BOOL);
+PMIX_INFO_REQUIRED(&directives[1]);
+```
+The first example instructs PMIx\_Log to generate a timestamp indicating when the log message was created, and to include that in the log. The second example requires that the log be written in XML format - an error must be returned if XML output support is not available.
+
+---
+
+**Advice to implementers:**  
+Calls relayed to the host thru the pmix\_server\_log\_fn\_t function provided in the pmix\_server\_module must _NOT_ call the PMIx\_Log API as this can result in an infinite loop. Instead, the implementation must perform one of three operations:
+ * transfer the data+directives to a "gateway" server where they can be logged. Gateways are designated servers on nodes (typically service nodes) where centralized logging is supported (e.g., writing to a global syslog). The data+directives may be passed to the PMIx\_Log API once arriving at that destination. If the PMIx server cannot support the provided request, it will call up to the host for final disposition - this is why the host must not call back into the server.
+
+ * transfer the data to a logging channel outside of PMIx, but directly supported by the host
+ 
+ * return an error to the caller indicating that the requested action is not supported
+
+A complete PMIx implementation will respond to PMIx\_Query\_info\_nb calls requesting logging channels supported by the RM with a comma-delimited string of channel keys - e.g., "PMIX\_LOG\_LOCAL\_SYSLOG,PMIX\_LOG\_EMAIL".
+
```

Comment:
This is fine for the RFC, but the standard, this is not "advice to implementers", this is "how it needs to work".

Reply:
I'm not sure I understand the concern here - the advice clearly states how a "complete implementation" must respond to the request. I don't see how that language would change when moving to a "standard" document. Could you perhaps give me an example of how you think it should read?

Reply:
This is ok.   I guess my question is:  if an implementation supports PMIx_Query and it supports logging and it supports PMIX_LOG_LOCAL_SYSLOG and PMIX_LOG_EMAIL, is it still optional for it to respond to a PMIx_Query_info call that requests feedback on which logging types are supported?   I guess I feel like major components (logging, querying, etc) might not be supported by an implementation, but if you are going to try to support something, returning unsupported for some aspects might be really hard for someone to code a pmix client around.   Oh this implementation supports logging, it supports query, but when I query about which types of logging it supports, it says, "Oh I don't support telling you which logging I support" is just really unfriendly.   We can say, "Well, what you have there is a crappy PMIx implementation", but at the same time it might be good to give the implementers better guidance that "these things" all work together and if you return "supported" for one of these features you really should return "supported for all of them.   I admit I'm unsure about this though.   I look at things like publish/lookup and there are types of publishing we don't support yet and I don't feel like we should return "unsupported" for all publish attempts just because we don't support some types of persistence settings or scope.   So I guess I'm really still torn on this whole thing.   Normally I would go with your idea that pretty much everything is a "quality of implementation" issue and any server can return "Not supported" on anything and if your client can't deal with that, take it up with your pmix server implementation.   However, I also know that right now we are in this state where we are trying to play catchup with the PMIx "standard" and probably the standard is advancing even faster than we are implementing.... so having at least some level of guidance between pmix server implementers and pmix client implementers on batches of functionality that should be delivered together would be helpful.   On the other hand, I know that we have a ton of work to do just to standardize what we have without the added burden of trying to group everything into bundles of functionality that should be implemented as a single unit.

Reply:
I hear the concern/confusion - others have raised similar concerns, and I sympathize with it. I also am unsure of the "right" solution, but I am beginning to lean towards providing more guidance. One reason for doing so is that we still have not seen/heard anyone indicate a plan to do an alternate implementation - all I hear is "I have no desire to write all that code".

So it might be possible to make stronger statements about the PMIx client/server requirements (basically requiring that everyone keep up with the ref implementation), and that would enable us to then perhaps provide a clearer base for the host RM to work against. It would also let us stabilize what the client sees as at least the PMIx client/server would be solid, so you could count on whatever support is in the PMIx server (leaving only the backend support as uncertain).

Probably something we need to discuss more thoroughly. I'll try to add it to the weekly agenda.